### PR TITLE
AIP-84 Migrate XCom get entries endpoint to Fastapi

### LINKED ISSUE #42980: AIP-84 Migrate XCom get entries public endpoint to FastAPI
### Description

Migrate the XCom get entries public endpoint from connexion to FastAPI

### Use case/motivation

[AIP-84](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-84+UI+REST+API)

### Related issues

Relates to: https://github.com/apache/airflow/issues/42370

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by michaeljs-c**
closes: #42980

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/api_fastapi/core_api/routes/public/xcom.py [hunk 1]
```diff
@@ -90,5 +93,56 @@ def get_xcom_entry(
...
+        filters=[],
+        order_by=SortParam(["dag_id", "task_id", "run_id", "map_index", "key"], XCom),
+        offset=offset,
+        limit=limit,
+        session=session,
+    )
+    xcoms = session.scalars(query)
+    return XComCollection(
+        xcom_entries=[XComResponse.model_validate(xcom) for xcom in xcoms], total_entries=total_entries
+    )
```
Comment (rawwar) [inline airflow/api_fastapi/core_api/routes/public/xcom.py]:
```suggestion
    return XComCollection(
        xcom_entries=xcoms, total_entries=total_entries
    )
```

File: airflow/api_fastapi/core_api/routes/public/xcom.py [hunk 2]
```diff
@@ -90,5 +93,56 @@ def get_xcom_entry(
...
+        query = query.where(XCom.task_id == task_id)
+    if dag_run_id != "~":
+        query = query.where(DR.run_id == dag_run_id)
+    if map_index is not None:
+        query = query.where(XCom.map_index == map_index)
+    if xcom_key is not None:
+        query = query.where(XCom.key == xcom_key)
+
+    query, total_entries = paginated_select(
+        select=query,
```
Comment (rawwar) [inline airflow/api_fastapi/core_api/routes/public/xcom.py]:
```suggestion
        statement=query,
```

Please pull changes from main. `select` has been renamed to `statement`

File: tests/api_fastapi/core_api/routes/public/test_xcom.py [hunk 3]
```diff
@@ -203,3 +208,279 @@ def test_custom_xcom_deserialize(
...
+                value=TEST_XCOM_VALUE,
+                run_id=run_id,
+                task_id=task_id,
+                dag_id=dag_id,
+                map_index=map_index,
+            )
+
+    @pytest.fixture(autouse=True)
+    def setup(self) -> None:
+        self.clear_db()
```
Comment (pierrejeambrun) [inline tests/api_fastapi/core_api/routes/public/test_xcom.py]: Nit: Can you move this setup/init code at the beginning of the test class please.

**Review by pierrejeambrun — COMMENTED** [review]
Nice.

Looking good, just one nitpick.

We can merge when @rawwar comments are addressed.

Thanks

File: airflow/api_fastapi/core_api/routes/public/xcom.py (hunk 2 above)
Comment (michaeljs-c) [reply airflow/api_fastapi/core_api/routes/public/xcom.py]: Thanks for highlighting, updated now.

File: tests/api_fastapi/core_api/routes/public/test_xcom.py (hunk 3 above)
Comment (michaeljs-c) [reply tests/api_fastapi/core_api/routes/public/test_xcom.py]: Yep makes sense! Done now

File: airflow/api_fastapi/core_api/routes/public/__init__.py [hunk 4]
```diff
@@ -68,10 +68,10 @@
 authenticated_router.include_router(plugins_router)
 authenticated_router.include_router(pools_router)
 authenticated_router.include_router(providers_router)
+authenticated_router.include_router(xcom_router)
 authenticated_router.include_router(task_instances_router)
 authenticated_router.include_router(tasks_router)
 authenticated_router.include_router(variables_router)
-authenticated_router.include_router(xcom_router)
```
Comment (amoghrajesh) [inline airflow/api_fastapi/core_api/routes/public/__init__.py:74]: nit: undo this change.
We are maintaining it in an alphabetical order

File: tests/api_fastapi/core_api/routes/public/test_xcom.py [hunk 5]
```diff
@@ -203,3 +208,279 @@ def test_custom_xcom_deserialize(
         else:
             assert response.status_code == 200
             assert response.json()["value"] == expected_status_or_value
+
+
+class TestGetXComEntries(TestXComEndpoint):
+    @pytest.fixture(autouse=True)
+    def setup(self) -> None:
+        self.clear_db()
```
Comment (amoghrajesh) [inline tests/api_fastapi/core_api/routes/public/test_xcom.py:216]: We shouldn't be needing this because setup and teardown are defined in class: `TestXComEndpoint` which `TestGetXComEntries` inherits from

**Review by amoghrajesh — COMMENTED** [review]
Looking nice @michaeljs-c! Two nits and we should be in a mergable state

File: airflow/api_fastapi/core_api/routes/public/__init__.py (hunk 4 above)
Comment (michaeljs-c) [reply airflow/api_fastapi/core_api/routes/public/__init__.py:74]: This was required due to a URL conflict with patch_task_instance endpoint:
- patch_task_instance: `"/{task_id}/{map_index}"`
- getEntries: `/{task_id}/getEntries`

Therefore it would throw a 422 error (because it tries to use "getEntries" as map_index which is defined as an int).

By moving xcom before task_instances, the xcom url is attempted first which resolves the issue, while the task instance endpoint is not affected.

Comment (amoghrajesh) [reply airflow/api_fastapi/core_api/routes/public/__init__.py:74]: Great thanks for the explanation.

File: tests/api_fastapi/core_api/routes/public/test_xcom.py (hunk 5 above)
Comment (michaeljs-c) [reply tests/api_fastapi/core_api/routes/public/test_xcom.py:216]: This was to override the original setup to not pre-include creating a dag run, since its handled slightly differently within `TestGetXComEntries._create_xcom_entries`

**Comment by michaeljs-c** [comment]
Looking into CI failures

**Comment by jason810496** [comment]
Hi @michaeljs-c, I think the current CI failure for static is cause by frontend generated code, which could be resolve by:
```
re-commit run ts-compile-format-lint-ui --all-files
```
refer: https://github.com/apache/airflow/pull/43859#issuecomment-2488570257

**Comment by jason810496** [comment]
> Looking into CI failures
Hi @michaeljs-c, I think the current CI failure for static is cause by frontend generated code, which could be resolve by:
```
pre-commit run ts-compile-format-lint-ui --all-files
```
refer: https://github.com/apache/airflow/pull/43859#issuecomment-2488570257

**Review by pierrejeambrun — APPROVED** [review]

**Review by amoghrajesh — APPROVED** [review]
Thanks for explanation. LGTM
